dvc/remote/gdrive/__init__.py

```diff
@@ -0,0 +1,519 @@
+        while retries < self.max_retries:
+            retries += 1
+            response = self.session.request(
+                method, self.GOOGLEAPIS_BASE_URL + path, *args, **kwargs
+            )
+            if (
+                self.response_is_ratelimit(response)
+                or response.status_code >= 500
+            ):
+                logger.debug(
+                    "got {} response, will retry in {} sec".format(
+                        response.status_code, exponential_backoff
+                    )
+                )
+                sleep(exponential_backoff)
+                exponential_backoff *= 2
+            else:
+                break
+        if response.status_code >= 400:
+            raise GDriveError(self.response_error_message(response))
+        return response
+
+    def get_metadata_by_id(self, file_id, **kwargs):
+        return self.request(
+            "GET", "drive/v3/files/" + file_id, **kwargs
+        ).json()
+
+    def search(self, parent=None, name=None, add_params={}):
+        query = []
+        if parent is not None:
+            query.append("'{}' in parents".format(parent))
+        if name is not None:
+            query.append("name = '{}'".format(name))
+        params = {"q": " and ".join(query), "spaces": self.space}
+        params.update(add_params)
+        while True:
+            data = self.request("GET", "drive/v3/files", params=params).json()
+            for i in data["files"]:
+                yield i
+            if not data.get("nextPageToken"):
+                break
+            params["pageToken"] = data["nextPageToken"]
+
+    def get_metadata_by_path(self, root, path, fields=[]):
+        parent = self.get_metadata_by_id(root)
+        current_path = ["gdrive://" + parent["id"]]
+        parts = path.split("/")
+        # only specify fields for the last search query
+        kwargs = [{}] * (len(parts) - 1) + [
+            {"add_params": {"fields": "files({})".format(",".join(fields))}}
+            if fields
+            else {}
+        ]
+        for part, kwargs in zip(parts, kwargs):
+            if not self.metadata_isdir(parent):
+                raise GDriveError(
+                    "{} is not a folder".format("/".join(current_path))
+                )
+            current_path.append(part)
+            files = list(self.search(parent["id"], part, **kwargs))
+            if len(files) > 1:
+                raise GDriveError(
+                    "path {} is duplicated".format("/".join(current_path))
+                )
+            elif len(files) == 0:
+                raise GDriveResourceNotFound("/".join(current_path))
+            parent = files[0]
+        return parent
+
+    def metadata_isdir(self, metadata):
+        return metadata["mimeType"] == self.MIME_GOOGLE_APPS_FOLDER
+
+    def get_file_checksum(self, path_info):
+        metadata = self.get_metadata_by_path(
+            path_info.root, path_info.path, params={"fields": "md5Checksum"}
```

Comment:
params  - unexpected argument

Reply:
Oops. Great catch! Thanks!

I wrote the test and it was failing, but `py.test` gave a bad assertion description which made me thought that this is a bug in `py.test` with the mock spec handling. It was probably late at night and I removed the spec (from this place and the all others! shame on me), so I ended up with the broken test here. I'll return the spec everywhere and file an issue for py.test assertion handling for spec'ed mocks.

Reply:
I focused on the tail lines describing the assertion and noticed the first `TypeError` line just now:

```
______________________________________________________________________________________________ test_get_file_checksum ______________________________________________________________________________________________
TypeError: got an unexpected keyword argument 'params'

The above exception was the direct cause of the following exception:

gdrive = RemoteGDrive: 'gdrive://root/data', monkeypatch = <_pytest.monkeypatch.MonkeyPatch object at 0x7f0490df2668>

    def test_get_file_checksum(gdrive, monkeypatch):
        monkeypatch.setattr(
            gdrive,
            "get_metadata_by_path",
            mock.Mock(
                gdrive.get_metadata_by_path,
                return_value=dict(id="id1", name="path1", md5Checksum="checksum")
            ),
        )
        checksum = gdrive.get_file_checksum(PathGDrive(gdrive.root, path="path1"))
        assert checksum == "checksum"
        gdrive.get_metadata_by_path.assert_called_once_with(
>           gdrive.root, "path1", params={"fields": "md5Checksum"}
        )

tests/unit/remote/test_gdrive.py:336: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
../../../../.virtualenvs/dvc/lib/python3.7/site-packages/mock/mock.py:957: in assert_called_once_with
    return self.assert_called_with(*args, **kwargs)
../../../../.virtualenvs/dvc/lib/python3.7/site-packages/mock/mock.py:944: in assert_called_with
    six.raise_from(AssertionError(_error_message(cause)), cause)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

value = None, from_value = TypeError("got an unexpected keyword argument 'params'")

>   ???
E   AssertionError: expected call not found.
E   Expected: mock('root', 'path1', params={'fields': 'md5Checksum'})
E   Actual: mock('root', 'path1', params={'fields': 'md5Checksum'})

<string>:3: AssertionError
```

Reply:
np! I hope pylint can detect things like this (when we enable it), in my case it was highlighted by pycharm :) so, not my catch even :)